src/partners/safello.ts

```diff
@@ -0,0 +1,99 @@
+import { asArray, asNumber, asObject, asString, asUnknown } from 'cleaners'
+import fetch from 'node-fetch'
+
+import { PartnerPlugin, PluginParams, PluginResult, StandardTx } from '../types'
+
+const asSafelloTx = asObject({
+  completedDate: asString,
+  id: asString,
+  currency: asString,
+  amount: asNumber,
+  cryptoCurrency: asString
+})
+
+const asSafelloResult = asObject({ orders: asArray(asUnknown) })
+const PER_REQUEST_LIMIT = 100
+const QUERY_LOOKBACK = 1000 * 60 * 60 * 24 * 5 // 5 days
+
+export async function querySafello(
+  pluginParams: PluginParams
+): Promise<PluginResult> {
+  const ssFormatTxs: StandardTx[] = []
+  let headers
+  let latestTimestamp = 0
+  if (typeof pluginParams.settings.lastCheckedTimestamp === 'number') {
+    latestTimestamp = pluginParams.settings.lastCheckedTimestamp
+  }
+
+  if (typeof pluginParams.apiKeys.apiKey === 'string') {
+    headers = {
+      'secret-key': pluginParams.apiKeys.apiKey
+    }
+  } else {
+    return {
+      settings: { lastCheckedTimestamp: latestTimestamp },
+      transactions: []
+    }
+  }
+
+  if (latestTimestamp > QUERY_LOOKBACK) {
+    latestTimestamp -= QUERY_LOOKBACK
+  }
+  let done = false
+  let offset = 0
+  let newestTimestamp = latestTimestamp
+  while (!done) {
+    const url = `https://app.safello.com/v1/partner/get-orders?offset=${offset}`
+    const result = await fetch(url, {
+      method: 'GET',
+      headers
+    })
+    const txs = asSafelloResult(await result.json())
+
+    for (const rawtx of txs.orders) {
+      const tx = asSafelloTx(rawtx)
+      const date = new Date(tx.completedDate)
+      const timestamp = date.getTime()
+      const ssTx = {
+        status: 'complete',
+        inputTXID: tx.id,
+        inputAddress: '',
+        inputCurrency: tx.currency,
+        inputAmount: tx.amount,
+        outputAddress: '',
+        outputCurrency: tx.cryptoCurrency,
+        outputAmount: tx.amount,
+        timestamp: timestamp / 1000,
+        isoDate: date.toISOString()
+      }
+      ssFormatTxs.push(ssTx)
+      if (timestamp > newestTimestamp) {
+        newestTimestamp = timestamp
+      }
+      if (latestTimestamp > timestamp) {
+        done = true
+      }
+    }
+
+    offset += PER_REQUEST_LIMIT
+
+    // reached end of database
+    if (txs.orders.length < PER_REQUEST_LIMIT) {
```

Comment:
UNDERSTANDING CHECK: not sure what this is doing? its going over the tx.orders and checking if the length is less than PER_REQUEST_LIMIT so if there are no more transactions left to check then tx.orders.length would be like 56 and the entire loop ends?